server/src/main/java/com/linecorp/centraldogma/server/internal/storage/DirectoryBasedStorageManager.java

```diff
@@ -323,6 +341,60 @@ public T unremove(String name) {
         return unremovedChild;
     }
 
+    @Override
+    public void markForPurge(String name) {
+        ensureOpen();
+        validateChildName(name);
+        File marked;
+        final File removed = new File(rootDir, name + SUFFIX_REMOVED);
+
+        final Supplier<File> newMarkedFile = () -> {
+            final String interfix = '.' + Long.toHexString(ThreadLocalRandom.current().nextLong());
+            return new File(rootDir, name + interfix + SUFFIX_PURGED);
+        };
+
+        synchronized (this) {
+            if (!removed.exists() || !removed.isDirectory()) {
+                throw newStorageNotFoundException(name + SUFFIX_REMOVED);
+            }
+
+            marked = newMarkedFile.get();
+            while (!removed.renameTo(marked)) {
+                marked = newMarkedFile.get();
+            }
+        }
+        final File purged = marked;
+        purgeWorker.execute(() -> deletePurgedFile(purged));
+    }
+
+    @Override
+    public void purgeMarked() {
+        ensureOpen();
+        final File[] files = rootDir.listFiles();
+        if (files == null) {
+            return;
+        }
+        for (File f : files) {
+            if (!f.isDirectory()) {
+                continue;
+            }
+            final String name = f.getName();
+            if (name.endsWith(SUFFIX_PURGED)) {
+                deletePurgedFile(f);
+            }
+        }
+    }
+
+    private void deletePurgedFile(File file) {
+        try {
+            logger.info("Deleting a purged {}: {} ..", childTypeName, file);
+            Util.deleteFileTree(file);
+            logger.info("Deleted a purged {}: {}.", childTypeName, file);
+        } catch (IOException e) {
+            logger.warn("Failed to delete a purged " + childTypeName + ": " + file, e);
```

Comment:
`"Failed to delete a purged {}: {}", childTypeName, file, e`

Reply:
same reason with https://github.com/line/centraldogma/pull/416#discussion_r305659075